path_dynamic_roles.go

```diff
@@ -0,0 +1,322 @@
+				},
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleRead,
+				},
+				logical.DeleteOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleDelete,
+				},
+			},
+			HelpSynopsis:    staticRoleHelpSynopsis,
+			HelpDescription: staticRoleHelpDescription,
+		},
+		// LIST role
+		{
+			Pattern: dynamicRolePath + "?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: b.pathDynamicRoleList,
+				},
+			},
+			HelpSynopsis:    "List all the dynamic roles Vault is currently managing in OpenLDAP.",
+			HelpDescription: "List all the dynamic roles being managed by Vault.",
+		},
+		// GET credentials
+		{
+			Pattern: path.Join(dynamicCredPath, framework.MatchAllRegex("name")),
+			Fields: map[string]*framework.FieldSchema{
+				"name": {
+					Type:        framework.TypeLowerCaseString,
+					Description: "Name of the dynamic role.",
+				},
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathDynamicCredsRead,
+				},
+			},
+			HelpSynopsis: "Request LDAP credentials for a dynamic role. These credentials are " +
+				"created within OpenLDAP when querying this endpoint.",
+			HelpDescription: "This path requests new LDAP credentials for a certain dynamic role. " +
+				"The credentials are created within OpenLDAP based on the creation_ldif specified " +
+				"within the dynamic role configuration.",
+		},
+	}
+}
+
+func dynamicSecretCreds(b *backend) *framework.Secret {
+	return &framework.Secret{
+		Type: secretCredsType,
+		Fields: map[string]*framework.FieldSchema{
+			"username": {
+				Type:        framework.TypeString,
+				Description: "Username of the generated account",
+			},
+			"password": {
+				Type:        framework.TypeString,
+				Description: "Password to access the generated account",
+			},
+			"distinguished_names": {
+				Type: framework.TypeStringSlice,
+				Description: "List of the distinguished names (DN) created. Each name in this list corresponds to" +
+					"each action taken within the creation_ldif statements. This does not de-duplicate entries, " +
+					"so this will have one entry for each LDIF statement within creation_ldif.",
+			},
+		},
+
+		Renew:  b.secretCredsRenew(),
+		Revoke: b.secretCredsRevoke(),
+	}
+}
+
+func (b *backend) pathDynamicRoleCreateUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	rawData := data.Raw
+	err := convertToDuration(rawData, "default_ttl", "max_ttl")
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert TTLs to duration: %w", err)
+	}
+
+	roleName := data.Get("name").(string)
+	dRole, err := retrieveDynamicRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, fmt.Errorf("unable to look for existing role: %w", err)
+	}
+	if dRole == nil {
+		dRole = &dynamicRole{}
+	}
```

Comment:
Does this mean that you can update a role that doesn't exist? I think that there needs to be a check here such that if this is an UpdateOperation but the role doesn't exist, we return an error. This is to prevent a token with `update` permission but not `create` permission from creating new roles.

Reply:
Ah good catch. Updated.